types.go

```diff
@@ -1781,10 +1783,12 @@ func NewMessage(data []byte) *Message {
 
 // done closes the internal doneSignal channel to let the receiver know that this message has been acted upon
 func (m *Message) done() {
-	// TODO: move initialization in ctor and use ctor everywhere?
-	if m.doneSignal != nil {
-		close(m.doneSignal)
```

Comment:
Why couldn't we just close the channel like before then set it to `nil`?  Is there concurrency at play?  From the description of the issue I thought it was just a double-disposition from the same goroutine.

Reply:
it could be concurrent yes, nothing prevents the receiver to spun concurrent go-routines on a message.

Reply:
To make sure I understand, you can have multiple goroutines receive the same message, and in this case, the first message disposition "wins"?

Reply:
You don't control what the user does. he is free to start as many goroutines he wants within his message handler.
At the moment, concurrent message disposition trigger a panic.
this occured in my tests (could be an Accept failure followed by an Abandon, for example)

The done channel frees up the linkcredit, it does not prevent subsequent dispositions to be sent. They will be sent as well, but from amqp linkcredit and message flow perspective, the message is considered "handled" when the first disposition is sent to the server.

At the moment, the link credit is incremented (listener side) when a disposition is sent. amqp does not look at the disposition result (success or error). could be a gap, but in practice, not really.

Reply:
Got it, this is happening in the user-defined message handler.  Makes sense now, thanks for the explanation.